lib/batch_verification/src/server.rs

```diff
@@ -0,0 +1,168 @@
+use tokio::{
+    io::AsyncWriteExt,
+    net::{TcpListener, TcpStream},
+};
+use tokio_util::codec::{FramedRead, FramedWrite};
+use zksync_os_l1_sender::batcher_model::BatchForSigning;
+use zksync_os_socket::skip_http_headers;
+
+/// Accepts connections from batch verification clients. Crafts and sends
+/// BatchVerificationRequests to all clients. Receives responses and forwards
+/// them through the channel to batch_response_processor
+pub struct BatchVerificationServer {
+    verification_request_broadcast: broadcast::Sender<BatchVerificationRequest>,
+    response_sender: mpsc::Sender<BatchVerificationResponse>,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[allow(clippy::large_enum_variant)]
+pub enum BatchVerificationRequestError {
+    #[error("Not enough clients connected")]
+    NotEnoughClients,
+    #[error("Failed to send batch verification request: {0}")]
+    SendError(#[from] broadcast::error::SendError<BatchVerificationRequest>),
+}
+
+impl BatchVerificationServer {
+    pub fn new() -> (Self, mpsc::Receiver<BatchVerificationResponse>) {
+        let (response_sender, response_receiver) = mpsc::channel(100);
+        let (verification_request_broadcast, _rx_unused) = broadcast::channel(16);
+
+        let server = Self {
+            verification_request_broadcast,
+            response_sender,
+        };
+
+        (server, response_receiver)
+    }
+
+    /// Start the TCP server that accepts connections from external nodes
+    pub async fn run_server(&self, address: impl ToSocketAddrs) -> anyhow::Result<()> {
+        let listener = TcpListener::bind(address).await?;
+        let response_sender = self.response_sender.clone();
+
+        loop {
+            let (socket, addr) = listener.accept().await?;
+            let verification_request_rx = self.verification_request_broadcast.subscribe();
+            let response_sender = response_sender.clone();
+            let client_addr = addr.to_string();
+
+            tokio::spawn(async move {
+                if let Err(e) = Self::handle_client(
+                    socket,
+                    client_addr,
+                    verification_request_rx,
+                    response_sender,
+                )
+                .await
+                {
+                    tracing::info!("Error handling client {}: {}", addr, e);
+                }
+            });
+        }
+    }
+
+    async fn handle_client(
+        mut socket: TcpStream,
+        client_addr: String,
+        mut verification_request_rx: broadcast::Receiver<BatchVerificationRequest>,
+        response_sender: mpsc::Sender<BatchVerificationResponse>,
+    ) -> anyhow::Result<()> {
+        let (recv, mut send) = socket.split();
+        let mut reader = BufReader::new(recv);
+
+        // Skip HTTP headers similar to replay_transport
+        skip_http_headers(&mut reader).await?;
+
+        // Write wire format version
+        send.write_u32(BATCH_VERIFICATION_WIRE_FORMAT_VERSION)
+            .await?;
```

Comment:
I am very concerned that the protocol implementation is 
1. Not encapsulated (we mix low-level protocol interchange with the component business logic).
2. Not documented (or at least I didn't find any docs).
3. Doesn't seem to be extendable (I worry about future compatibility somewhat).

If we want to take a look at Ethereum, it already has an established communication protocol: [rlpx](https://github.com/ethereum/devp2p/blob/master/rlpx.md). Moreover, it's extendable: you can add new capabilities and implement sub-protocols easily for different purposes.
If we take a look at engineering globally, we have protobuf/capnp/etc, which would help with a lot of stuff we currently have to worry about.

Was it a deliberate/discussed decision to implement a new protocol here?